### ISSUE TITLE
Option to add good upscaling for the mask images

### ISSUE THREAD

**Issue by ghost** [Mikubill/sd-webui-controlnet#20]
When I use higher resolutions the black and white mask image gets blurry and the details wont come out as good.

Can it be upscaled in the background with SwinIR or Esrgan, when choosing a resolution above 512x512? The details would then stay reasonably crispy

**Comment by Mikubill** [comment]
Nice idea! But two questions:
1. Are original models trained by lllyasviel tough enough to support such large resolution?
2. Currently we used `F.interpolate("nearest")` to handle Hires.Fix stuffs, not a good method but works. Could SwinIR supports such op without extra costs?
https://github.com/Mikubill/sd-webui-controlnet/blob/fca33937af7bf45dac94616531643ce228c88d6b/scripts/cldm.py#L379-L383

After cleaning up runtime issues, I will take a look on it.

**Comment by corebro** [comment]
I found SwinIR4 to be the best, but it takes a bit processing to get there.

Esrgan is sufficient and its faster. I dont know if it would be better to upscale the generated mask image, or to upscale the picture first and then let the model generate in that resolution.

Also, I read that the controlnet models can be a lot smaller but I dont remember how exactly. Someone mentioned only a small part is needed that is inside the model file and this would lower vram further. Has to be extracted

And so far I know interpolate nearest is very fast but it is blurry. One pass with Esrgan how it could be made with the highres fix would be optimal, I guess. Im not a very talented python programmer otherwise I would have probably contributed

**Comment by corebro** [comment]
![Annotation 2023-02-14 091345](https://user-images.githubusercontent.com/125164139/218677441-599b63c9-806b-4b49-b7dc-fd6984962307.png)

There are some artifacts in the mask images too that lead to sharp coarse edges. Could be the model
